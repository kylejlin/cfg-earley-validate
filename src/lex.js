const lex = (rules, string) => {
  const tokens = [];
  let line = 1;
  let column = 0;
  while (string.length > 0) {
    for (const rule of rules) {
      const [regex, tokenizer] = rule;
      const match = string.match(regex);
      if (match === null || match.index !== 0 || match[0] === '') {
        continue;
      }
      const [firstMatch] = match;
      
      string = string.slice(firstMatch.length);

      const startLine = line;
      const startColumn = column;
      const newlines = firstMatch.split('\n').length - 1;
      if (newlines > 0) {
        line += newlines;
        column = firstMatch.split('\n').slice(-1)[0].length;
      } else {
        column += firstMatch.length;
      }

      if (tokenizer === null) {
        break;
      }
      if ('string' === typeof tokenizer) {
        tokens.push({
          type: tokenizer,
          value: firstMatch,
          start: [startLine, startColumn],
          end: [line, column],
        });
        break;
      }
      if ('function' === typeof tokenizer) {
        tokens.push(tokenizer({
          type: null,
          value: firstMatch,
          start: [startLine, startColumn],
          end: [line, column],
        }));
        break;
      }
      throw new TypeError('Illegal tokenizer: ' + tokenizer);
    }
  }
  return tokens;
};

export default lex;
