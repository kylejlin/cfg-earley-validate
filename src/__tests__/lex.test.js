import lex from '../lex';

test('it lexes correctly', () => {
  const rules = [
    [
      /\(/,
      'LPAREN'
    ],
    [
      /\)/,
      'RPAREN'
    ],
    [
      /\d*(?:\.\d+)?/,
      (token) => ({
        ...token,
        type: 'NUMBER',
        value: parseFloat(token.value),
      })
    ],
    [
      /"(?:[^"\\]|\\.)*"/,
      (token) => ({
        ...token,
        type: 'STRING',
        value: token.value.slice(1, -1),
      })
    ],
    [
      /[a-zA-Z_][a-zA-Z_0-9]*/,
      'IDENTIFIER'
    ],
    [
      /[ \t\n]+/,
      null
    ]
  ];
  const code = '(foo 42 (baz "bar"))';
  expect(lex(rules, code)).toMatchSnapshot();
});
