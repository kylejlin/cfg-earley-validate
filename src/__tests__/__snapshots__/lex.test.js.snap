// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it lexes correctly 1`] = `
Array [
  Object {
    "end": Array [
      1,
      1,
    ],
    "start": Array [
      1,
      0,
    ],
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "end": Array [
      1,
      4,
    ],
    "start": Array [
      1,
      1,
    ],
    "type": "IDENTIFIER",
    "value": "foo",
  },
  Object {
    "end": Array [
      1,
      7,
    ],
    "start": Array [
      1,
      5,
    ],
    "type": "NUMBER",
    "value": 42,
  },
  Object {
    "end": Array [
      1,
      9,
    ],
    "start": Array [
      1,
      8,
    ],
    "type": "LPAREN",
    "value": "(",
  },
  Object {
    "end": Array [
      1,
      12,
    ],
    "start": Array [
      1,
      9,
    ],
    "type": "IDENTIFIER",
    "value": "baz",
  },
  Object {
    "end": Array [
      1,
      18,
    ],
    "start": Array [
      1,
      13,
    ],
    "type": "STRING",
    "value": "bar",
  },
  Object {
    "end": Array [
      1,
      19,
    ],
    "start": Array [
      1,
      18,
    ],
    "type": "RPAREN",
    "value": ")",
  },
  Object {
    "end": Array [
      1,
      20,
    ],
    "start": Array [
      1,
      19,
    ],
    "type": "RPAREN",
    "value": ")",
  },
]
`;
